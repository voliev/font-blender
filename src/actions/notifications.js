import { ADD_NOTIFICATION, REMOVE_NOTIFICATION } from "../constants";

export const addNotification = (text, group) => ({
  type: ADD_NOTIFICATION,
  text,
  group
});

export const removeNotification = group => ({
  type: REMOVE_NOTIFICATION,
  group
});
