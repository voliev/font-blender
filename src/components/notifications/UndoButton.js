import React from "react";
import PropTypes from "prop-types";
import "./styles/UndoButton.css";

const UndoButton = ({ onClickHandler }) => (
  <button
    className="UndoButton btn"
    onClick={onClickHandler}
    value="undo"
    title="Undo"
  >
    Undo
  </button>
);

UndoButton.propTypes = {
  onClickHandler: PropTypes.func.isRequired
};

export default UndoButton;
