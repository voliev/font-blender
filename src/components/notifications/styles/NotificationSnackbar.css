.NotificationSnackbar {
  align-items: center;
  background-color: var(--black);
  border-radius: var(--border-radius-md);
  color: var(--white);
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  max-height: 4.25rem;
  min-height: 3rem;
  padding: var(--space-sm) var(--space-xs) var(--space-sm) var(--space-md);
  transition: all 0.4s;
  width: 21.5rem;
}

.NotificationSnackbar:not(:last-child) {
  margin-bottom: var(--space-xs);
}

.NotificationSnackbar .notification {
  align-items: center;
  display: flex;
  font-size: var(--font-xs);
  margin: 0 var(--space-xs) 0 0;
}

.NotificationSnackbar .controls {
  align-items: center;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
}

.NotificationSnackbar .close-icon {
  color: var(--white);
  transition: transform 0.1s;
}

.NotificationSnackbar .close-icon:hover {
  transform: scale(1.2);
}

.NotificationSnackbar.elevate {
  animation: elevate 0.4s forwards;
}

.NotificationSnackbar.drop,
.NotificationSnackbar.undo {
  animation: fadeOut 0.2s forwards;
}
