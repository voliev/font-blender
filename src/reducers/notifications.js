import { ADD_NOTIFICATION, REMOVE_NOTIFICATION } from "../constants";

const groups = ["selectedFonts", "styles", "clipboard"];

const initialState = groups.reduce((res, group) => {
  res[group] = {
    current: "",
    pending: ""
  };
  return res;
}, {});

export default (state = initialState, action) => {
  switch (action.type) {
    case ADD_NOTIFICATION:
      return {
        ...state,
        [action.group]: {
          current: state[action.group].pending
            ? state[action.group].pending
            : state[action.group].current || action.text,
          pending: !state[action.group].current ? "" : action.text
        }
      };
    case REMOVE_NOTIFICATION:
      return {
        ...state,
        [action.group]: {
          current: state[action.group].pending,
          pending: ""
        }
      };
    default:
      return state;
  }
};
