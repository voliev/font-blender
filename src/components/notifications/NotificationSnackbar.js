import React, { PureComponent } from "react";
import classNames from "classnames";
import ClearOutlined from "@material-ui/icons/ClearOutlined";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { TIMEOUT } from "../../constants";
import UndoButton from "./UndoButton";
import "./styles/NotificationSnackbar.css";

export class NotificationSnackbar extends PureComponent {
  constructor() {
    super();

    this.state = {
      stage: "elevate",
      timerId: undefined
    };
    this.handleAnimationEnd = this.handleAnimationEnd.bind(this);
    this.handleUndoButtonClick = this.handleUndoButtonClick.bind(this);
    this.handleCloseIconClick = this.handleCloseIconClick.bind(this);
  }

  handleAnimationEnd() {
    const { group, removeNotification, undoActionHandler } = this.props;

    const handlers = {
      elevate: () => this.setTimer(),
      drop: removeNotification,
      undo: undoActionHandler
    };
    const { stage } = this.state;

    if (this.state.stage === "drop") {
      handlers[stage](group);
    } else {
      handlers[stage]();
    }
  }

  handleUndoButtonClick() {
    this.clearTimer("undo");
  }

  handleCloseIconClick() {
    this.clearTimer("drop");
  }

  setTimer() {
    const timerId = setTimeout(() => {
      this.clearTimer("drop");
    }, TIMEOUT);

    this.setState({
      stage: "display",
      timerId
    });
  }

  clearTimer(stage) {
    clearTimeout(this.state.timerId);

    this.setState({
      stage,
      timerId: undefined
    });
  }

  componentDidUpdate(prevProps) {
    const { stage } = this.state;
    const { pending } = this.props.notification;
    if (stage === "display" && pending) {
      this.clearTimer("drop");
    }

    // Group's state change triggers notification stage change
    if (this.props.groupState) {
      const { groupState } = this.props;
      const { groupState: prevGroupState } = prevProps;
      const isStateWentForward = groupState.length > prevGroupState.length;
      if (isStateWentForward && this.state.stage === "display") {
        this.clearTimer("drop");
      }
    }
  }

  render() {
    const { current: notification } = this.props.notification;
    const { stage } = this.state;
    const componentClasses = classNames({
      NotificationSnackbar: true,
      [stage]: true
    });

    return (
      <div
        className={componentClasses}
        onAnimationEnd={this.handleAnimationEnd}
      >
        <p className="notification">{notification}</p>
        <div className="controls">
          {this.props.undoActionHandler ? (
            <UndoButton onClickHandler={this.handleUndoButtonClick} />
          ) : null}
          <button
            className="btn btn-round-small close-icon"
            onClick={this.handleCloseIconClick}
            title="Close"
            value="close"
          >
            <ClearOutlined fontSize="small" />
          </button>
        </div>
      </div>
    );
  }
}

NotificationSnackbar.propTypes = {
  notification: PropTypes.exact({
    current: PropTypes.string.isRequired,
    pending: PropTypes.string.isRequired
  }).isRequired,
  group: PropTypes.string.isRequired,
  removeNotification: PropTypes.func.isRequired,
  undoActionHandler: PropTypes.func,
  groupState: PropTypes.any
};

const mapStateToProps = (state, ownProps) => ({
  groupState:
    state[ownProps.group] && "present" in state[ownProps.group]
      ? state[ownProps.group].present
      : undefined
});

export default connect(mapStateToProps)(NotificationSnackbar);
