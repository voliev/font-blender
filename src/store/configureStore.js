import { createStore, combineReducers, applyMiddleware, compose } from "redux";
import thunk from "redux-thunk";
import fonts from "../reducers/fonts";
import filters from "../reducers/filters";
import selectedFonts from "../reducers/selectedFonts";
import showcase from "../reducers/showcase";
import styles from "../reducers/styles";
import notifications from "../reducers/notifications";

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export default () => {
  const store = createStore(
    combineReducers({
      fonts,
      filters,
      selectedFonts,
      showcase,
      styles,
      notifications
    }),
    composeEnhancers(applyMiddleware(thunk))
  );

  return store;
};
