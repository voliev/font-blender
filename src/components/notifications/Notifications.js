import React from "react";
import PropTypes from "prop-types";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import NotificationSnackbar from "./NotificationSnackbar";
import { startUndoDeselected } from "../../actions/selectedFonts";
import { startUndoReset } from "../../actions/styles";
import { removeNotification } from "../../actions/notifications";
import "./styles/Notifications.css";

export const Notifications = ({
  notifications,
  removeNotification,
  ...rest
}) => {
  return (
    <div className="Notifications">
      {Object.keys(notifications).map(group =>
        notifications[group].current ? (
          <NotificationSnackbar
            key={notifications[group].current}
            group={group}
            notification={notifications[group]}
            undoActionHandler={rest[group]}
            removeNotification={removeNotification}
          />
        ) : null
      )}
    </div>
  );
};

Notifications.propTypes = {
  notifications: PropTypes.objectOf(
    PropTypes.exact({
      current: PropTypes.string.isRequired,
      pending: PropTypes.string.isRequired
    })
  ).isRequired,
  removeNotification: PropTypes.func.isRequired,
  selectedFonts: PropTypes.func.isRequired,
  styles: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  notifications: state.notifications
});

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      selectedFonts: startUndoDeselected,
      styles: startUndoReset,
      removeNotification
    },
    dispatch
  );

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Notifications);
