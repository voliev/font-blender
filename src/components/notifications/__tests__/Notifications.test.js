import React from "react";
import { mount, shallow } from "enzyme";
import { Provider } from "react-redux";
import thunk from "redux-thunk";
import configureStore from "redux-mock-store";
import NotificationsConnected, { Notifications } from "../Notifications";
import NotificationSnackbar from "../NotificationSnackbar";
import {
  REMOVE_NOTIFICATION,
  UNDO_DESELECTED,
  UNDO_RESET
} from "../../../constants";
import notifications from "../../../fixtures/notifications";
import state from "../../../fixtures/state";

const setup = (
  renderFn = shallow,
  propsOverrides = {},
  connected = false,
  provideStore = false
) => {
  const props = {
    notifications: {
      ...notifications,
      selectedFonts: {
        ...notifications.selectedFonts,
        current: "Roboto font family has been removed"
      },
      styles: {
        ...notifications.styles,
        current: "Styles has been reset"
      }
    },
    removeNotification: jest.fn(),
    selectedFonts: jest.fn(),
    styles: jest.fn(),
    ...propsOverrides
  };

  if (connected || provideStore) {
    const mockStore = configureStore([thunk]);
    const store = mockStore(state);

    return {
      ...props,
      store,
      wrapper: connected
        ? renderFn(<NotificationsConnected store={store} />)
        : renderFn(
            <Provider store={store}>
              <Notifications {...props} />
            </Provider>
          )
    };
  }

  return {
    ...props,
    wrapper: renderFn(<Notifications {...props} />)
  };
};

it("should render without crashing", () => {
  const { wrapper } = setup();

  expect(wrapper).toMatchSnapshot();
});

it("should receive corresponding props", () => {
  const {
    notifications,
    removeNotification,
    selectedFonts,
    styles,
    wrapper
  } = setup(mount, undefined, false, true);

  expect(typeof wrapper.find(Notifications).prop("notifications")).toBe(
    "object"
  );
  expect(wrapper.find(Notifications).prop("notifications")).toEqual(
    notifications
  );

  expect(typeof wrapper.find(Notifications).prop("removeNotification")).toBe(
    "function"
  );
  expect(wrapper.find(Notifications).prop("removeNotification")).toEqual(
    removeNotification
  );

  expect(typeof wrapper.find(Notifications).prop("selectedFonts")).toBe(
    "function"
  );
  expect(wrapper.find(Notifications).prop("selectedFonts")).toEqual(
    selectedFonts
  );

  expect(typeof wrapper.find(Notifications).prop("styles")).toBe("function");
  expect(wrapper.find(Notifications).prop("styles")).toEqual(styles);
});

it("should render notification snackbar if there are no notifications", () => {
  const { wrapper } = setup(shallow, { notifications });

  expect(wrapper.find(NotificationSnackbar).length).toBe(0);
});

it("should render notification snackbar if there are current notifications", () => {
  const { wrapper } = setup();

  expect(wrapper.find(NotificationSnackbar).length).toBe(2);
});

it("should dispatch actions", () => {
  const { wrapper, store } = setup(shallow, undefined, true);

  wrapper.prop("selectedFonts")();

  expect(store.getActions()).toContainEqual({
    type: UNDO_DESELECTED
  });
  expect(store.getActions()).toContainEqual({
    type: REMOVE_NOTIFICATION,
    group: "selectedFonts"
  });

  store.clearActions();

  expect(store.getActions()).toEqual([]);

  wrapper.prop("styles")();

  expect(store.getActions()).toContainEqual({
    type: UNDO_RESET
  });
  expect(store.getActions()).toContainEqual({
    type: REMOVE_NOTIFICATION,
    group: "styles"
  });

  store.clearActions();

  expect(store.getActions()).toEqual([]);

  wrapper.prop("removeNotification")("styles");

  expect(store.getActions()).toContainEqual({
    type: REMOVE_NOTIFICATION,
    group: "styles"
  });
});
