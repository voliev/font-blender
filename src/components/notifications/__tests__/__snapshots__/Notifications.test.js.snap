// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render without crashing 1`] = `
<div
  className="Notifications"
>
  <Connect(NotificationSnackbar)
    group="selectedFonts"
    key="Roboto font family has been removed"
    notification={
      Object {
        "current": "Roboto font family has been removed",
        "pending": "",
      }
    }
    removeNotification={[MockFunction]}
    undoActionHandler={[MockFunction]}
  />
  <Connect(NotificationSnackbar)
    group="styles"
    key="Styles has been reset"
    notification={
      Object {
        "current": "Styles has been reset",
        "pending": "",
      }
    }
    removeNotification={[MockFunction]}
    undoActionHandler={[MockFunction]}
  />
</div>
`;
